game-headless/src/main/java/org/triplea/game/server/debug/ChatHandler.java

```diff
@@ -69,7 +71,7 @@ private static void sendChatMessage(final String message) {
         .ifPresent(chat -> chat.sendMessage(message));
   }
 
-  private String formatChatMessage(final LogRecord record) {
-    return getFormatter().format(record).trim();
+  private List<String> formatChatMessage(final LogRecord record) {
+    return Arrays.asList(getFormatter().format(record).trim().split("\n"));
```

Comment:
Yeah, if it triggers Error Prone, it should actually fail the build because of this line:

https://github.com/triplea-game/triplea/blob/789bcf2d0a20ad16138dd96f12ddf17ee400f0be/build.gradle#L125

Since the build passed, I have a feeling the following line:

https://github.com/triplea-game/triplea/blob/789bcf2d0a20ad16138dd96f12ddf17ee400f0be/build.gradle#L134

might be overwriting the preceding Error Prone configuration and effectively disabling everything.  Regardless, line 134 should probably be removed and replaced with a `disable 'Finally'` in the prior `options.errorprone` closure.

Reply:
Hmm, on one hand, the API wants a regex, so '\n' might be weird. Beyond the manual verification, I added a test case to further verify newline splitting and changed the "\n" to a proper regex "\\n".

Reply:
Yeah, but error-prone should really not let the build pass.
Have a look at this: https://github.com/google/error-prone/blob/master/docs/bugpattern/StringSplit.md
That's why this error-prone pattern exists

Reply:
Hmm, I'm not sure I see the problem. I removed the suppression line in the build, 134, and the build still passed. I don't understand exactly what that config is doing to be confident in removing it.

`split(String)` is equivalent to: `split(String, int)`, the 'zero' value seems like the behavior we would want (ie: "\n\n\n" would be split to an empty array):

```
     *    <li><p>
     *    If the <i>limit</i> is zero then the pattern will be applied as
     *    many times as possible, the array can have any length, and trailing
     *    empty strings will be discarded.</p></li>
     *
     *    <li><p>
     *    If the <i>limit</i> is negative then the pattern will be applied
     *    as many times as possible and the array can have any length.</p></li>
```

- Can we restore this error prone rule?
- Is it appropriate we take this up in this PR? Considering we have other usages of split in other places.
- Will this actually come back to bite us in any way, what is the behavior that we are concerned about? In this case specifically, and I suppose in general. Given that problem or statement of concern, is it actually worth us taking our time to use Guava splitter, as compared to say trying to finish JavaFX, deprecating the Java NIO server, or any other functionality that we can build? I personally appreciate the small marginal gains that add up over time, but if we're going to disallow `String.split(String)`, let's get the error prone rule working, or add some strong notes about it in the wiki convention, or let's not worry about it if it's only a problem in the abstract.

Reply:
> ie: "\n\n\n" would be split to an empty array

This might be the thing most people don't expect. I would expect it to result in an array of 4 empty strings.
`"\n\n\na".split("\n");` on the other hand does result in an array of 4, 3 empty strings, 1 string `"a"`, but `"a\n\n\n".split("\n");` results in an array with a single element `"a"`.
Also if `"\n\n"` and `"\n"` split into an empty array, why does splitting `""` return an array with a single empty string?
This is all weird an super unintuitive, which is why error-prone suggest replacing the usage with something more explicit.

Anyways the main problem right now is that error-prone doesn't cause the build to fail even though it should.
We invested effort in the past to remove all usages of normal `String#split(String)` in order to be able to make this rule error out the build rather than just a warning.

It would be nice if @ssoloff could confirm or bust his assumption to be able to avoid having such discussions in the future.

Reply:
If removing line 134 didn't cause the build to fail, I'd say my theory is incorrect.

There may be something more fundamental going on here.  I'd recommend trying the following:

* Was Error Prone upgraded recently?  If so, you may want to check the release notes to see if there were any changes in the `StringSplitter` rule (e.g. maybe it is now context-sensitive rather than being triggered on simply any usage of `String#split(String)`).
* Is the `StringSplitter` rule dependent on the JDK version?  All the `StringSplitter` violations were fixed in the Java 8 build.  Maybe there's something about the Java 11+ builds that prevents it from triggering?
* Introduce some code that was known to trigger `StringSplitter` in the past (e.g. #4074) and verify the build fails (or at least the violation is reported to the console).

If that last suggestion does not trigger a violation, you may want to create an MCVE that attempts to trigger the rule without all the TripleA baggage.  If you can't trigger it in an MCVE, it's probably worthy of opening an Error Prone issue to find out what's going on.  If you can trigger it, then you at least know it's something with the TripleA build environment, and you can begin peeling that onion.

Reply:
I managed to use the Web UI to try a few one-line changes (see [this branch](https://github.com/ssoloff/triplea-game-triplea/commits/string-splitter-test)).

It appears `StringSplitter` does not simply trigger on the presence of `String#split(String)`, but rather it's affected by whether or not the result of `split()` is assigned to a `String[]`.  For example, [this change](https://github.com/ssoloff/triplea-game-triplea/commit/e6064a1219e56aeb57fc9023c09f7248049a7de9) triggers the rule:

```diff
diff --git a/game-headless/src/main/java/org/triplea/game/server/debug/ChatHandler.java b/game-headless/src/main/java/org/triplea/game/server/debug/ChatHandler.java
index 30908cdfd9..31d79567de 100644
--- a/game-headless/src/main/java/org/triplea/game/server/debug/ChatHandler.java
+++ b/game-headless/src/main/java/org/triplea/game/server/debug/ChatHandler.java
@@ -70,6 +70,7 @@ private static void sendChatMessage(final String message) {
   }
 
   private String formatChatMessage(final LogRecord record) {
+    final String[] temp = getFormatter().format(record).trim().split("\n");
     return getFormatter().format(record).trim();
   }
 }
```

But [this one](https://github.com/ssoloff/triplea-game-triplea/commit/81a2d0f6e678b08a4f898b6998c25aac0b5944ce) does not:

```diff
diff --git a/game-headless/src/main/java/org/triplea/game/server/debug/ChatHandler.java b/game-headless/src/main/java/org/triplea/game/server/debug/ChatHandler.java
index 31d79567de..6e4c1b43ac 100644
--- a/game-headless/src/main/java/org/triplea/game/server/debug/ChatHandler.java
+++ b/game-headless/src/main/java/org/triplea/game/server/debug/ChatHandler.java
@@ -70,7 +70,7 @@ private static void sendChatMessage(final String message) {
   }
 
   private String formatChatMessage(final LogRecord record) {
-    final String[] temp = getFormatter().format(record).trim().split("\n");
+    getFormatter().format(record).trim().split("\n");
     return getFormatter().format(record).trim();
   }
 }
```

That would explain why the usage in this PR doesn't trigger--because the return value is being passed directly to `Arrays#asList()` and not assigned to a `String[]`.

Not sure if this is a defect in `StringSplitter` or if it's by design.  That would probably require spending some time examining the [implementation](https://github.com/google/error-prone/blob/master/core/src/main/java/com/google/errorprone/bugpatterns/StringSplitter.java) and the corresponding [tests](https://github.com/google/error-prone/blob/master/core/src/test/java/com/google/errorprone/bugpatterns/StringSplitterTest.java).